vowpalwabbit/core/tests/io_alignment_test.cc

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "vw/core/io_buf.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+using namespace testing;
+
+using align_t = VW::desired_align::align_t;
+using flatbuffer_t = VW::desired_align::flatbuffer_t;
+using VW::desired_align;
+
+struct positioned_ptr
+{
+  size_t allocation;
+  void* allocation_unit;
+  int8_t* p;
+  static_assert(sizeof(int8_t) == 1, "int8_t is not 1 byte");
+
+  positioned_ptr(size_t allocation)
+      : allocation(allocation), allocation_unit(malloc(allocation)), p(reinterpret_cast<int8_t*>(allocation_unit))
+  {
+  }
+  positioned_ptr(const positioned_ptr&) = delete;
+  positioned_ptr(positioned_ptr&& original)
+      : allocation(original.allocation), allocation_unit(original.allocation_unit), p(original.p)
+  {
+    original.allocation_unit = nullptr;
+    original.allocation = 0;
+    original.p = nullptr;
+  }
+
+  ~positioned_ptr()
+  {
+    if (allocation_unit != nullptr) { free(allocation_unit); }
+  }
+
+  void realign(align_t alignment, align_t offset)
+  {
+    size_t base_address = reinterpret_cast<size_t>(allocation_unit);
+    size_t base_offset = base_address % alignment;
+
+    size_t padding = alignment - base_offset + offset;
+    assert(padding < allocation);
+
+    p += padding;
+  }
+};
```

Comment:
```suggestion
  /* Changes p so that it is proper alignment.  For example if alignment is 4
     and p will point to the a location in allocation_unit that has the right   alignment */
  void realign(align_t alignment, align_t offset)
```